alacritty/src/migrate/mod.rs

```diff
@@ -0,0 +1,286 @@
+//! Configuration file migration.
+
+use std::fmt::Debug;
+use std::path::Path;
+use std::{fs, mem};
+
+use tempfile::NamedTempFile;
+use toml_edit::{DocumentMut, Item};
+
+use crate::cli::MigrateOptions;
+use crate::config;
+
+mod yaml;
+
+/// Handle migration.
+pub fn migrate(options: MigrateOptions) {
+    // Find configuration file path.
+    let config_path = options
+        .config_file
+        .clone()
+        .or_else(|| config::installed_config("toml"))
+        .or_else(|| config::installed_config("yml"));
+
+    // Abort if system has no installed configuration.
+    let config_path = match config_path {
+        Some(config_path) => config_path,
+        None => {
+            eprintln!("No configuration file found");
+            std::process::exit(1);
+        },
+    };
+
+    // If we're doing a wet run, perform a dry run first for safety.
+    if !options.dry_run {
+        #[allow(clippy::redundant_clone)]
+        let mut options = options.clone();
+        options.silent = true;
+        options.dry_run = true;
+        if let Err(err) = migrate_config(&options, &config_path, config::IMPORT_RECURSION_LIMIT) {
+            eprintln!("Configuration file migration failed:");
+            eprintln!("    {config_path:?}: {err}");
+            std::process::exit(1);
+        }
+    }
+
+    // Migrate the root config.
+    match migrate_config(&options, &config_path, config::IMPORT_RECURSION_LIMIT) {
+        Ok(migration) => {
+            if !options.silent {
+                println!("{}", migration.success_message(false));
+            }
+        },
+        Err(err) => {
+            eprintln!("Configuration file migration failed:");
+            eprintln!("    {config_path:?}: {err}");
+            std::process::exit(1);
+        },
+    }
+}
+
+/// Migrate a specific configuration file.
+fn migrate_config<'a>(
+    options: &MigrateOptions,
+    path: &'a Path,
+    recursion_limit: usize,
+) -> Result<Migration<'a>, String> {
+    // Ensure configuration file has an extension.
+    let path_str = path.to_string_lossy();
+    let (prefix, suffix) = match path_str.rsplit_once('.') {
+        Some((prefix, suffix)) => (prefix, suffix),
+        None => return Err("missing file extension".to_string()),
+    };
+
+    // Handle legacy YAML files.
+    if suffix == "yml" {
+        let new_path = yaml::migrate(options, path, recursion_limit, prefix)?;
+        return Ok(Migration::Yaml((path, new_path)));
+    }
+
+    // TOML only does renames, so return early if they are disabled.
+    if options.skip_renames {
+        return Ok(Migration::Toml(path));
+    }
```

Comment:
I think we should indicate error here, so it's less confusing.

Reply:
It's possible to have migrations that have both yaml and toml files though. So it's not technically an error and there might be other things done by migrate in the future which aren't renames.

That said, I doubt this option was ever used outside of testing. So I'll settle for a warning for now and we can probably remove it together with the yaml stuff.